aten/src/ATen/core/boxing/impl/boxing.h

```diff
@@ -7,73 +7,252 @@
 #include <c10/core/TensorOptions.h>
 #include <ATen/core/boxing/KernelFunction.h>
 
-namespace at {
-struct Dimname;
-}
+#include <ATen/core/Dimname.h>
 
 namespace c10 {
 namespace impl {
 
+//
+// utils
+//
+
+// is_specialization_of
+//
+template <template <typename...> class T, typename U>
+struct is_specialization_of: std::false_type {};
+
+template <template <typename...> class T, typename... Us>
+struct is_specialization_of<T, T<Us...>>: std::true_type {};
+
+// is_tuple_of_lvalue_refs
+//
+template<class T, bool = is_specialization_of<std::tuple, T>::value>
+struct is_tuple_of_lvalue_refs :
+  guts::typelist::all<std::is_lvalue_reference, guts::typelist::from_tuple_t<T>>
+{};
+
+template<class T>
+struct is_tuple_of_lvalue_refs<T, false> : std::false_type {};
+
+// tuple_elements
+//
+template <class Tuple, size_t... ns>
+constexpr auto tuple_elements(Tuple t, std::index_sequence<ns...>) {
+    return std::tuple<std::tuple_element_t<ns, Tuple>...>(std::get<ns>(t)...);
+}
+
+// tuple_take
+//
+template <class Tuple, size_t n>
+constexpr auto tuple_take(Tuple t) {
+    return tuple_elements(t, std::make_index_sequence<n>{});
+}
+
+//
+// boxing predicates
+//
+
+// A boxable arg type is one that IValue has a constructor for.
 // Assume T is decayed
 template <typename T>
-using not_ok_to_box = guts::negation<guts::disjunction<
+using can_box =
+  guts::disjunction<
     std::is_constructible<IValue, T>,
     // TensorOptions are not directly constructible into IValue,
     // but torch::jit::push knows how to handle them
-    std::is_same<TensorOptions, T>,
-    // void returns are ok
-    std::is_same<void, T>>>;
+    std::is_same<TensorOptions, T>
+  >;
+
+template <typename... Ts>
+using can_box_all = guts::conjunction<can_box<std::decay_t<Ts>>...>;
+
+// an unboxable result is one that can be extracted from an IValue
+template <typename T>
+using can_unbox =
+  guts::conjunction<
+    guts::disjunction<
+      // using IValue(T) as a proxy for IValue.to<T>() here
```

Comment:
you could also check for IValue.to<T> directly by creating a type trait with `void_t`.

```cpp
template<class T, class Enable = void> struct has_ivalue_to : std::false_type {};
template<class T> struct has_ivalue_to<T, guts::void_t<decltype(IValue::to<T>())>> : std::true_type {};
```

Reply:
ran into an unexpected wrinkle here, leaving it as is for now but will follow up